cluster-autoscaler/core/scaledown/latencytracker/node_latency_tracker_test.go

```diff
@@ -0,0 +1,208 @@
+		unneededList     []string
+		scaledDownList   []string
+		unremovableList  []string
+		wantTrackedNodes []string
+		wantThresholds   map[string]time.Duration
+	}{
+		{
+			name:             "add new unneeded nodes",
+			setupNodes:       []string{"node1"},
+			unneededList:     []string{"node1", "node2", "node3"},
+			scaledDownList:   []string{},
+			unremovableList:  []string{},
+			wantTrackedNodes: []string{"node1", "node2", "node3"},
+		},
+		{
+			name:             "node is scaled down",
+			setupNodes:       []string{},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{"node1"},
+			unremovableList:  []string{},
+			wantTrackedNodes: []string{},
+		},
+		{
+			name:             "node becomes needed",
+			setupNodes:       []string{},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{},
+			unremovableList:  []string{"node1"},
+			wantTrackedNodes: []string{},
+		},
+		{
+			name:             "node is still unneeded",
+			setupNodes:       []string{"node1"},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{},
+			unremovableList:  []string{},
+			wantTrackedNodes: []string{"node1"},
+		},
+		{
+			name: "mix of different nodes",
+			setupNodes: []string{
+				"node1",
+				"node2",
+				"node3",
+				"node4",
+			},
+			unneededList:     []string{"node1", "node4", "node5", "node6"},
+			scaledDownList:   []string{"node4"},
+			unremovableList:  []string{"node6"},
+			wantTrackedNodes: []string{"node1", "node5"},
+		},
+		{
+			name:             "process node that was never tracked",
+			setupNodes:       []string{},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{"node2"},
+			unremovableList:  []string{"node3"},
+			wantTrackedNodes: []string{"node1"},
+		},
+		{
+			name:             "update threshold for known node",
+			setupNodes:       []string{},
+			updateThresholds: map[string]time.Duration{"node1": 4 * time.Second},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{},
+			unremovableList:  []string{},
+			wantTrackedNodes: []string{"node1"},
+			wantThresholds:   map[string]time.Duration{"node1": 4 * time.Second},
+		},
+		{
+			name:             "update threshold for unknown node",
+			updateThresholds: map[string]time.Duration{"node2": 5 * time.Second},
+			unneededList:     []string{"node1"},
+			scaledDownList:   []string{},
+			unremovableList:  []string{},
+			wantTrackedNodes: []string{"node1"},
+			wantThresholds:   map[string]time.Duration{"node1": 0 * time.Second},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tracker := NewNodeLatencyTracker(processor.NewDefaultScaleDownStatusProcessor())
+
+			if len(tt.setupNodes) > 0 {
+				tracker.UpdateScaleDownCandidates(nodesFromNames(tt.setupNodes), baseTime)
+			}
+
+			tracker.UpdateScaleDownCandidates(nodesFromNames(tt.unneededList), baseTime.Add(5*time.Second))
+
+			for node, threshold := range tt.updateThresholds {
```

Comment:
In actual code `UpdateThreshold` will actually be called first since it is called within scale down planner, while `UpdateScaleDownCandidates` is called after the list of unneeded nodes is ready.